Add Cellbase DB Restoration to Mongo cluster creation

lawrencegripper opened the pull request:
After a successful setup of Mongo on Node 0 in the Mongo cluster the machine will download the Cellbase Dump defined in the top level template variable and restore it to the cluster. 

(I also fixed a formatting errors in some files which is why they show changes)

j-coll commented:
@lawrencegripper
If we only have the MongoDB with the cellbase data, we'll need to modify the storage-configuration file.
In case that we add the the REST server, we'll need also to modify the IVA configuration.

```yaml
cellbase.database.hosts: <mongodb-nodes>
cellbase.database.user: <username>
cellbase.database.password: <password>
cellbase.database.options.authenticationDatabase: admin
cellbase.database.options.sslEnabled: true

storageEngines[*].variant.options.annotator: "cellbase_db_adaptor"
```

Example:
```yaml
---
cellbase:
  hosts:      ## REST connection (to be modified when we have the REST server)
  - "http://bioinfo.hpc.cam.ac.uk/cellbase/"
  version: "v4"
  database:   ## CellBase MongoDB direct connection
    hosts:
    - "opencgamongo01:27017"
    - "opencgamongo02:27017"
    user: "mongouser"
    password: "mongopassword"
    options:
      authenticationDatabase: "admin"
      sslEnabled : true

...

storageEngines:
 - id : "mongodb"
   variant:
     options:
       ....
       annotator: "cellbase_db_adaptor"
       ....
 - id : "hadoop"
   variant:
     options:
       ....
       annotator: "cellbase_db_adaptor"
       ....
```

lawrencegripper commented:
Hopefully done all those config changes in the `init` container. I'm going to wait for the build to finish then do a deployment with these images. 

What is the best way to test that this is working correctly?

lawrencegripper commented:
So this is good to go in its current form. I've added a script extension to the VM which will check that the `cloud-init` applies correctly as it was driving me nuts doing it manually.

lawrencegripper commented:
Looks like this work for the expected for the normal config but errors due to the IVA configuration. Just updating now to fix that issue.

martinpeck commented on opencga-app/app/scripts/docker/opencga-init/override-yaml.py:
```diff
@@ -1,8 +1,10 @@
 import oyaml as yaml
 import argparse
 import sys
+import os
```
probably don't need this

lawrencegripper replied:
Good spot should be fixed.

martinpeck commented on opencga-app/app/scripts/docker/opencga-init/Dockerfile:
```diff
@@ -30,7 +30,7 @@ ARG IVA_TAG="v1.0.0-rc1"
 # Install local dependencies
 RUN apt-get update && apt-get install -y python3 python3-pip git && \
-    pip3 install --upgrade pip setuptools && \
+    pip3 install --upgrade pip setuptools ConfigArgParse && \
```
might consider a requirements.txt instead

lawrencegripper replied:
Agree, now fixed.

lawrencegripper commented:
Think I'm on the home straight now. I've:

-  Simplified the way that params get passed into the `override-yaml.py` script to reduce the duplication and hopefully simplify edits in future. 
- Untangled the massive line that started the init container in the daemonVM template into a multiline `concat` statement. Editors flag this as invalid but ARM is happy with an ARM function split over multiple lines. 
- Added more tests for the configuration
- Moved config tests to work off the current builds config files so config changes which break the script will flag during docker builds 
- Updated the default config to include the missing healthCheck value by default
- Moved to `jsonc` extension for the `DaemonVM` template and added comments. This seems to play nicely!!!!!

lawrencegripper commented:
One last test run tomorrow and I'll be all done on this one.

lawrencegripper commented:
This is good to go! Wooop!

Requires the following params to test. Would be good to get a review and have someone test out a deployment. 

> Note: The commit SHA on the PR isn't the same as the image SHA used as Jenkins appears to be merging the PRs with the release-1.4.1 branch before building which results in a different SHA for the docker images. 

```

        "_artifactsLocation": {
            "value": "https://raw.githubusercontent.com/lawrencegripper/opencga/lg/issue/1105-dbrestore/opencga-app/app/scripts/azure/arm/"
        },
        "batchContainerImage": {
            "value": "opencb/opencga-batch:dda9fcba4529f8cab535060726d14c1919898e4f"
        },
        "openCGAContainerImage": {
            "value": "opencb/opencga-app:dda9fcba4529f8cab535060726d14c1919898e4f"
        },
        "ivaContainerImage": {
            "value": "opencb/iva:v1.0.0-rc2"
        },
        "initContainerImage": {
            "value": "opencb/opencga-init:dda9fcba4529f8cab535060726d14c1919898e4f"
        },
        "daemonContainerImage": {
            "value": "opencb/opencga-daemon:dda9fcba4529f8cab535060726d14c1919898e4f"
        },
```

lawrencegripper commented:
Found another bug in the Hadoop configuration setup in the YAML

marrobi commented:
@lawrencegripper great will give it a test, might not have time today given time needed to restore cellbase, but will see.  I need to use the above docker parameters and the cellbase ones in the updated tempaltes?

marrobi commented on opencga-app/app/scripts/azure/arm/azuredeploy.json:
```diff
@@ -501,7 +521,7 @@
                 "mode": "Incremental",
                 "templateLink": {
-                    "uri": "[concat(parameters('_artifactsLocation'), '/daemonvm/azuredeploy.json', parameters('_artifactsLocationSasToken'))]",
+                    "uri": "[concat(parameters('_artifactsLocation'), '/daemonvm/azuredeploy.jsonc', parameters('_artifactsLocationSasToken'))]",
```
What was the outcome of the jsonc experiment? Have we got some jsonc and some json?

lawrencegripper replied:
Yup I've moved the Deamon VM to `.jsonc` as was testing and all works fine. Don't know why the ARM team didn't mention that.

lawrencegripper commented:
Ready for test now with new SHA

```

        "_artifactsLocation": {
            "value": "https://raw.githubusercontent.com/lawrencegripper/opencga/lg/issue/1105-dbrestore/opencga-app/app/scripts/azure/arm/"
        },
        "batchContainerImage": {
            "value": "opencb/opencga-batch:e7ba3a682a8156e72234e9e0b27609298b95991a"
        },
        "openCGAContainerImage": {
            "value": "opencb/opencga-app:e7ba3a682a8156e72234e9e0b27609298b95991a"
        },
        "ivaContainerImage": {
            "value": "opencb/iva:v1.0.0-rc2"
        },
        "initContainerImage": {
            "value": "opencb/opencga-init:e7ba3a682a8156e72234e9e0b27609298b95991a"
        },
        "daemonContainerImage": {
            "value": "opencb/opencga-daemon:e7ba3a682a8156e72234e9e0b27609298b95991a"
        },
```

marrobi commented on opencga-app/app/scripts/azure/arm/azuredeploy.parameters.json:
```diff
@@ -50,6 +50,9 @@
         },
+        "mongoDbDumpUrl": {
+            "value": ""
+        },
```
Don't think this is needed any more?
```suggestion

```

lawrencegripper replied:
Good spot I'll remove this one! Thanks.

marrobi commented on opencga-app/app/scripts/docker/opencga/Dockerfile:
```diff
@@ -13,8 +13,9 @@ ARG buildPath="./build"
+ENV OPENCGA_CONFIG_DIR=/opt/opencga/conf
 
-RUN adduser -u 1001 opencga
+RUN useradd -u 1001 opencga
```
Add -m to create home, verified:

```root@7bac65164994:/opt/opencga# ls /home
opencga```

```suggestion
RUN useradd -u 1001 -m opencga
```

lawrencegripper replied:
Added thanks!

marrobi commented:
@lawrencegripper so this all seems to be good as far as I can see, providing the job log in slack looks OK (other than timings) to @j-coll .  Suggest get it merged, if anything needs fixing up later it can be. Great job.

lawrencegripper commented:
Agreed don't want this one getting any bigger. Will merge now.
